Add blog post Introduction to Deep RL (Deep RL Class Unit 1) (#285)
huggingface/blog: deeprlclass-introduction-to-deep-reinforcement-learning.md

<figure class="image table text-center m-0 w-full">
  <img src="assets/63_deeprlclass_unit1/OpenAIFive.jpg" alt="OpenAI Five, an AI that beat some of the best Dota2 players in the world"/>
  <figcaption>OpenAI Five, an AI <a href="https://www.twitch.tv/videos/293517383">that beat some of the best Dota2 players in the world</a></figcaption>
</figure>

Moreover, since 2018, **you have now, access to so many amazing environments and libraries to build your agents.**

That’s why this is the best moment to start learning, and with this course **you’re in the right place.**

Yes, because this article is the first unit of [Deep Reinforcement Learning Class](https://github.com/huggingface/deep-rl-class), a **free class from beginner to expert** where you’ll learn the theory and practice using famous Deep RL libraries such as Stable Baselines3, RL Baselines3 Zoo and RLlib.

In this free course, you will:

- 📖 Study Deep Reinforcement Learning in **theory and practice**.

> - 🧑‍💻 Learn to **use famous Deep RL libraries* such as Stable Baselines3, RL Baselines3 Zoo, and RLlib.

Reviewer: ```suggestion
- 🧑‍💻 Learn to **use famous Deep RL libraries** such as Stable Baselines3, RL Baselines3 Zoo, and RLlib.
```

> - 🤖 Train agents in **unique environments** such as SnowballFight, Huggy the Doggo 🐶, and classical ones such as Space Invaders and PyBullet.

Reviewer: Should we add a link to the snowball fight space to get people excited?

Author: Sure, I add it

- 💾 Publish your trained agents **in one line of code to the Hub**. But also download powerful agents from the community.
- 🏆 **Participate in challenges** where you will evaluate your agents against other teams.
- 🖌️🎨 Learn to **share your environments** made with Unity and Godot.

> So in this first unit, **you’ll learn the foundations of deep reinforcement learning.** And then, you'll train your first lander agent to **land correctly on the Moon 🌕 and upload it to the 🤗 Hub.**

Reviewer: ```suggestion
So in this first unit, **you’ll learn the foundations of deep reinforcement learning.** And then, you'll train your first lander agent to **land correctly on the Moon 🌕 and upload it to the Hugging Face Hub, a free, open platform where people can share ML models, datasets and demos.**
```

<figure class="image table text-center m-0 w-full">
    <video
        alt="LunarLander"
        style="max-width: 70%; margin: auto;"
        autoplay loop autobuffer muted playsinline
    >
      <source src="assets/63_deeprlclass_unit1/lunarlander.mp4" type="video/mp4">
  </video>
</figure>

It’s essential **to master these elements** before diving into implementing Deep Reinforcement Learning agents. The goal of this chapter is to give you solid foundations.

If you prefer, you can watch the 📹 video version of this chapter :

> <iframe width="560" height="315" src="https://www.youtube.com/embed/q0BiUn5LiBc?start=127" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

Reviewer: The video link starts at start=127. Is that intended?

Author: Yes since the introduction speaks about Deep RL v2

> So let’s get started! 🚀

Reviewer: Maybe a ToC such as in https://huggingface.co/blog/summer-at-huggingface would make sense

Author: :heavy_check_mark:

## **What is Reinforcement Learning?**

> To understand reinforcement learning, let’s start with the big picture.

Reviewer: nit: I would suggest to consistently capitalize throughout the blog. In some places you say Reinforcement Learning and in others you say reinforcement learning

Author: :heavy_check_mark:

  <img src="assets/63_deeprlclass_unit1/Illustration_1.jpg" alt="Illustration_1"/>
</figure>

Your brother will interact with the environment (the video game) by pressing the right button (action). He got a coin, that’s a +1 reward. It’s positive, he just understood that in this game **he must get the coins.**

<figure class="image table text-center m-0 w-full">
  <img src="assets/63_deeprlclass_unit1/Illustration_2.jpg" alt="Illustration_2"/>
</figure>

But then, **he presses right again** and he touches an enemy, he just died -1 reward.

<figure class="image table text-center m-0 w-full">
  <img src="assets/63_deeprlclass_unit1/Illustration_3.jpg" alt="Illustration_3"/>
</figure>

By interacting with his environment through trial and error, your little brother understood that **he needed to get coins in this environment but avoid the enemies.**

**Without any supervision**, the child will get better and better at playing the game.

That’s how humans and animals learn, **through interaction.** Reinforcement Learning is just a **computational approach of learning from action.**

### **A formal definition**

If we take now a formal definition:

> > Reinforcement learning is a framework for solving control tasks (also called decision problems) by building agents that learn from the environment by interacting with it through trial and error and receiving rewards (positive or negative) as unique feedback.

Reviewer: Where does this quote/definition come from?

Author: Don't remember :O I searched, it's not from Sutton, not from OpenAI, I think I've done a mix between multiple definition.

>

⇒ But how Reinforcement Learning works?

## **The Reinforcement Learning Framework**
### **The RL Process**

<figure class="image table text-center m-0 w-full">

>   <img src="assets/63_deeprlclass_unit1/RL_process.jpg" alt="The RL process"/>

Reviewer: Is this image yours? If not, let's link to source

Author: Its mine but it comes from Sutton I add the source

</figure>

- Our Agent receives **state S0** from the **Environment** — we receive the first frame of our game (Environment).
- Based on that **state S0,** the Agent takes **action A0** — our Agent will move to the right.
- Environment goes to a **new** **state S1** — new frame.
- The environment gives some **reward R1** to the Agent — we’re not dead *(Positive Reward +1)*.

This RL loop outputs a sequence of **state, action, reward and next state.**

<figure class="image table text-center m-0 w-full">
  <img src="assets/63_deeprlclass_unit1/sars.jpg" alt="State, Action, Reward, Next State"/>
</figure>

The agent's goal is to maximize its cumulative reward, **called the expected return.**

### **The reward hypothesis: the central idea of Reinforcement Learning**

⇒ Why is the goal of the agent to maximize the expected return?

Because RL is based on the **reward hypothesis**, which is that all goals can be described as the **maximization of the expected return** (expected cumulative reward).

That’s why in Reinforcement Learning, **to have the best behavior,** we need to **maximize the expected cumulative reward.**

> ### **(Optional) Markov Property**

Reviewer: ```suggestion
### **Markov Property**
```

In papers, you’ll see that the RL process is called the **Markov Decision Process** (MDP).

> We’ll talk again about the Markov Property in the following units. But if you need to remember something today about it, Markov Property implies that our agent needs **only the current state to decide** what action to take and **not the history of all the states** **and actions** he took before.

Reviewer: ```suggestion
We’ll talk again about the Markov Property in the following units. But if you need to remember something today about it, Markov Property implies that our agent needs **only the current state to decide** what action to take and **not the history of all the states** **and actions** they took before.
```

  <figcaption>A Self Driving Car agent has an infinite number of possible actions since he can turn left 20°, 21°, 22°, honk, turn right 20°, 20,1°…
</figcaption>
</figure>

To recap:
<figure class="image table text-center m-0 w-full">
  <img src="assets/63_deeprlclass_unit1/action_space.jpg" alt="Recap action space"/>
</figcaption>
</figure>

Taking this information into consideration is crucial because it will **have importance when choosing the RL algorithm in the future.**

### **Rewards and the discounting**

The reward is fundamental in RL because it’s **the only feedback** for the agent. Thanks to it, our agent knows **if the action taken was good or not.**

The cumulative reward at each time step t can be written as:

<figure class="image table text-center m-0 w-full">
  <img src="assets/63_deeprlclass_unit1/rewards_1.jpg" alt="Rewards"/>
  <figcaption>The cumulative reward equals to the sum of all rewards of the sequence.
</figcaption>
</figure>

Which is equivalent to:

<figure class="image table text-center m-0 w-full">
  <img src="assets/63_deeprlclass_unit1/rewards_2.jpg" alt="Rewards"/>

> </figure>

Reviewer: For non math people, let's maybe say something along these lines

The above equation is adding the reward at a given moment 0, plus the reward at the moment 1, and so on.

Author: :heavy_check_mark:

> However, in reality, **we can’t just add them like that.** The rewards that come sooner (at the beginning of the game) **are more likely to happen** since they are more predictable than the long-term future reward.

Reviewer: Would it make sense to mention that if the agent finds a positive return and there is no discount, it will just keep doing the same thing since that will maximize reward?

Author: Not really, in fact it will depends on his policy, if for instance we define an epsilon greedy policy (for instance we Q or DQN). Our agent will exploit 1-epsilon probability.

In other terms, how to build an RL agent that can **select the actions that maximize its expected cumulative reward?**

### **The Policy π: the agent’s brain**

The Policy **π** is the **brain of our Agent**, it’s the function that tell us what **action to take given the state we are.** So it **defines the agent’s behavior** at a given time.

<figure class="image table text-center m-0 w-full">
  <img src="assets/63_deeprlclass_unit1/policy_1.jpg" alt="Policy"/>
  <figcaption>Think of policy as the brain of our agent, the function that will tells us the action to take given a state
</figcaption>
</figure>

Think of policy as the brain of our agent, the function that will tells us the action to take given a state

This Policy **is the function we want to learn**, our goal is to find the optimal policy **π*, the policy that** maximizes **expected return** when the agent acts according to it. We find this **π* through training.**

There are two approaches to train our agent to find this optimal policy π*:

- **Directly,** by teaching the agent to learn which **action to take,** given the state is in: **Policy-Based Methods.**

> - Indirectly, **teach the agent to learn which state is more valuable** and then take the action that **leads to the more valuable states**: Value-Based Methods.

Reviewer: Best explanation of value based ever!

Author: Thanks :smile:

### **Policy-Based Methods**

In Policy-Based Methods, **we learn a policy function directly.**

This function will map from each state to the best corresponding action at that state. **Or a probability distribution over the set of possible actions at that state.**

<figure class="image table text-center m-0 w-full">
  <img src="assets/63_deeprlclass_unit1/policy_2.jpg" alt="Policy"/>
  <figcaption>As we can see here, the policy (deterministic) <b>directly indicates the action to take for each step.</b>
</figcaption>
</figure>

> As we can see here, the policy (deterministic) **directly indicates the action to take for each step.**

Reviewer: THis is repeated from caption

Author: :heavy_check_mark:

If we recap:

<figure class="image table text-center m-0 w-full">
  <img src="assets/63_deeprlclass_unit1/vbm_1.jpg" alt="Vbm recap"/>
</figure>
<figure class="image table text-center m-0 w-full">
  <img src="assets/63_deeprlclass_unit1/vbm_2.jpg" alt="Vbm recap"/>
</figure>

## **The “Deep” in Reinforcement Learning**

⇒ Wait… you spoke about Reinforcement Learning, but why we spoke about Reinforcement Learning?

Deep Reinforcement Learning introduces **deep neural networks to solve Reinforcement Learning problems** — hence the name “deep.”

For instance, in the next article, we’ll work on Q-Learning (classic Reinforcement Learning) and then Deep Q-Learning both are value-based RL algorithms.

You’ll see the difference is that in the first approach, **we use a traditional algorithm** to create a Q table that helps us find what action to take for each state.

In the second approach, **we will use a Neural Network** (to approximate the q value).

<figure class="image table text-center m-0 w-full">
  <img src="assets/63_deeprlclass_unit1/deep.jpg" alt="Value based RL"/>
  <figcaption>Schema inspired by the Q learning notebook by Udacity
</figcaption>
</figure>

> If you are not familiar with Deep Learning you definitely should watch <a href="http://introtodeeplearning.com/">the MIT Intro Course on Deep Learning (Free)</a>

Reviewer: Maybe fastai is a better intro resources https://course.fast.ai/

Author: Sure! I change it :heavy_check_mark:

- The RL process is a loop that outputs a sequence of **state, action, reward and next state.**
- To calculate the expected cumulative reward (expected return), we discount the rewards: the rewards that come sooner (at the beginning of the game) **are more probable to happen since they are more predictable than the long term future reward.**
- To solve an RL problem, you want to **find an optimal policy**, the policy is the “brain” of your AI that will tell us **what action to take given a state.** The optimal one is the one who **gives you the actions that max the expected return.**

- There are two ways to find your optimal policy:
  1. By training your policy directly: **policy-based methods.**
  2. By training a value function that tells us the expected return the agent will get at each state and use this function to define our policy: **value-based methods.**

- Finally, we speak about Deep RL because we introduces **deep neural networks to estimate the action to take (policy-based) or to estimate the value of a state (value-based)** hence the name “deep.”

> 

Reviewer: Maybe let's say something that they should not worry if they did not understand all the terms/concepts, in the next few weeks they will be diving into the topic, reading more resources and taking the learnings into practical things

Reviewer: Actually I saw you put something similar later :smile: maybe let's merge those two

---

> Now that you've studied the bases of Reinforcement Learning, you’re ready to train your first lander agent to **land correctly on the Moon 🌕 and upload it to the 🤗 Hub** 🔥

Reviewer: ```suggestion
Now that you've studied the bases of Reinforcement Learning, you’re ready to train your first lander agent to **land correctly on the Moon 🌕 and share it with the community through the Hub** 🔥
```

<figure class="image table text-center m-0 w-full">
    <video
        alt="LunarLander"
        style="max-width: 70%; margin: auto;"
        autoplay loop autobuffer muted playsinline
    >
      <source src="assets/63_deeprlclass_unit1/lunarlander.mp4" type="video/mp4">
  </video>
</figure>

> Start the tutorial here [ADD LINK]

Reviewer: TODO before merge
